src/utils/get-global-config.js

```diff
@@ -0,0 +1,36 @@
+const Configstore = require('configstore')
+const { v4: uuidv4 } = require('uuid')
+
+const { rmFileAsync, readFileAsync } = require('../lib/fs')
+const { getPathInHome, getLegacyPathInHome } = require('../lib/settings')
+
+const globalConfigDefaults = {
+  /* disable stats from being sent to Netlify */
+  telemetryDisabled: false,
+  /* cliId */
+  cliId: uuidv4(),
+}
+
+const getGlobalConfig = async function () {
```

Comment:
[sand] Not sure there is big benefit for it, but would it make sense to use `lodash.once` to resolve the config only a single time.
For example in `utils/command.js`  the function will be invoked twice (once for the token and once for the whole config).

Reply:
Yep! Good one, I actually tested it out locally and saw some specifc commands could actually trigger a call to `globalConfig` multiple times, I wanted to mention that but just forgot 🤦  I think some kind of memoisation like `lodash.once` would definitely be a good thing here 👍

Reply:
Small suggestion: [`memoize-one`](https://github.com/alexreardon/memoize-one) is also an option.
As opposed to `lodash.once()`, it [is tiny](https://github.com/alexreardon/memoize-one/blob/master/src/memoize-one.ts).

`lodash.once()`:

```js
var FUNC_ERROR_TEXT = 'Expected a function';

var INFINITY = 1 / 0,
    MAX_INTEGER = 1.7976931348623157e+308,
    NAN = 0 / 0;

var symbolTag = '[object Symbol]';

var reTrim = /^\s+|\s+$/g;

var reIsBadHex = /^[-+]0x[0-9a-f]+$/i;

var reIsBinary = /^0b[01]+$/i;

var reIsOctal = /^0o[0-7]+$/i;

var freeParseInt = parseInt;

var objectProto = Object.prototype;

var objectToString = objectProto.toString;

function before(n, func) {
  var result;
  if (typeof func != 'function') {
    throw new TypeError(FUNC_ERROR_TEXT);
  }
  n = toInteger(n);
  return function() {
    if (--n > 0) {
      result = func.apply(this, arguments);
    }
    if (n <= 1) {
      func = undefined;
    }
    return result;
  };
}

function once(func) {
  return before(2, func);
}

function isObject(value) {
  var type = typeof value;
  return !!value && (type == 'object' || type == 'function');
}

function isObjectLike(value) {
  return !!value && typeof value == 'object';
}

function isSymbol(value) {
  return typeof value == 'symbol' ||
    (isObjectLike(value) && objectToString.call(value) == symbolTag);
}

function toFinite(value) {
  if (!value) {
    return value === 0 ? value : 0;
  }
  value = toNumber(value);
  if (value === INFINITY || value === -INFINITY) {
    var sign = (value < 0 ? -1 : 1);
    return sign * MAX_INTEGER;
  }
  return value === value ? value : 0;
}

function toInteger(value) {
  var result = toFinite(value),
      remainder = result % 1;

  return result === result ? (remainder ? result - remainder : result) : 0;
}

function toNumber(value) {
  if (typeof value == 'number') {
    return value;
  }
  if (isSymbol(value)) {
    return NAN;
  }
  if (isObject(value)) {
    var other = typeof value.valueOf == 'function' ? value.valueOf() : value;
    value = isObject(other) ? (other + '') : other;
  }
  if (typeof value != 'string') {
    return value === 0 ? value : +value;
  }
  value = value.replace(reTrim, '');
  var isBinary = reIsBinary.test(value);
  return (isBinary || reIsOctal.test(value))
    ? freeParseInt(value.slice(2), isBinary ? 2 : 8)
    : (reIsBadHex.test(value) ? NAN : +value);
}

module.exports = once;
```

`memoize-one`:

```js
  function areInputsEqual(newInputs, lastInputs) {
      if (newInputs.length !== lastInputs.length) {
          return false;
      }
      for (var i = 0; i < newInputs.length; i++) {
          if (newInputs[i] !== lastInputs[i]) {
              return false;
          }
      }
      return true;
  }

  function memoizeOne(resultFn, isEqual) {
      if (isEqual === void 0) { isEqual = areInputsEqual; }
      var lastThis;
      var lastArgs = [];
      var lastResult;
      var calledOnce = false;
      function memoized() {
          var newArgs = [];
          for (var _i = 0; _i < arguments.length; _i++) {
              newArgs[_i] = arguments[_i];
          }
          if (calledOnce && lastThis === this && isEqual(newArgs, lastArgs)) {
              return lastResult;
          }
          lastResult = resultFn.apply(this, newArgs);
          calledOnce = true;
          lastThis = this;
          lastArgs = newArgs;
          return lastResult;
      }
      return memoized;
  }
```

Reply:
Good point @ehmicky! So I was actually looking at `lodash.once` and was considering alternatives, however seems like we already have lodash as a `dependency` - https://github.com/netlify/cli/blob/master/package.json#L132 - taking that into account would it be worth bringing another dependency? Might be something worth considering if we plan on decoupling `lodash` from the project (or at least shift to depending only on the specific lodash modules we need).

Reply:
Good point!
Either works, your call :+1:

Reply:
> however seems like we already have lodash as a `dependency`

That's the main reason I suggested it :)

See [here](https://github.com/netlify/cli/pull/1551) about using specific modules.

It would be great to dump `lodash` all together in favor of specific packages. I'll open an issue for it.

Reply:
> It would be great to dump lodash all together in favor of specific packages. I'll open an issue for it.

If we do want to go this direction, then using a non-lodash package might be a good idea, since this would remove the need to migrate that specific package.

Reply:
I'm using `memoizeOne` now, removed the legacy file deletion and fine tuned a couple of things. Let me know what you think @ehmicky @erezrokah 👍

Reply:
Given we're keeping the "legacy" config path for now, would it be worth it creating an issue just to track this "tech debt"? CC @erezrokah

Reply:
Looks good! :+1: 
About adding an issue for future dead code removal: this would be a good idea.

Quick question: was there anyway to always retrieve the global configuration at the beginning of each command (in other words: at the beginning of the base command constructor) and assign it to `this` so that it is available to every command? This would remove the need for calling `getGlobalConfig()` in each command, and also the need to memoize.

Reply:
Good catch @ehmicky. I guess it would be possible and it would probably make a lot of sense for the `track/identify` telemetry functions to receive the required config via its parameters instead of requiring it. For the hooks however, I'm still digging through oclifs arch but seems like they don't have access to the base command instance - https://oclif.io/docs/hooks#lifecycle-events 😕. We also have this static `getToken` function that depends on the global-config - https://github.com/netlify/cli/blob/master/src/utils/command.js#L313 - but seems like it's only being called on the deploy test - https://github.com/netlify/cli/blob/master/tests/command.deploy.test.js#L8 (?)

Tbh I think it makes sense to avoid this memoisation 🤔 maybe eventually remove it. Just not 100% of what would be the impact of doing it right now, given the global-config seems to be tied to a couple of different places.

Reply:
We could move forward with `memoisation` and open another issue to refactor the code to use `this.netlify.globalConfig`.
WDYT?

Reply:
Sounds good! :+1: